substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/thread/VMThreadSTFeature.java

```diff
@@ -173,14 +173,14 @@ public boolean apply(GraphBuilderContext b, ResolvedJavaMethod targetMethod, Rec
     private boolean handleGet(GraphBuilderContext b, ResolvedJavaMethod targetMethod, Receiver receiver) {
         VMThreadLocalInfo info = threadLocalCollector.findInfo(b, receiver.get());
         VMThreadLocalSTHolderNode holder = b.add(new VMThreadLocalSTHolderNode(info));
-        b.addPush(targetMethod.getSignature().getReturnKind(), new LoadVMThreadLocalNode(b.getMetaAccess(), info, holder, BarrierType.PRECISE));
+        b.addPush(targetMethod.getSignature().getReturnKind(), new LoadVMThreadLocalNode(b.getMetaAccess(), info, holder, BarrierType.UNKNOWN));
```

Comment:
The correct new barrier type to use in this class (both changes) is `ARRAY`: In Substrate VM's single threaded mode, the thread local variables are stored in a normal Java array that resides in the image heap.

Reply:
This should probably be an implementation detail of LoadVMThreadLocalNode instead of being a constructor parameter.

Reply:
I don't think so. VMThreadMTFeature.java passes NONE there. I'll change it to ARRAY for the VMThreadSTFeature.

Reply:
Changed to ARRAY in latest push.

Reply:
The difference between the `MT` and `ST` feature is correct: in multi-threaded (MT) mode, the thread local variables are in the thread descriptor is in C memory, i.e., no barriers needed. In single-threaded (ST) mode, the thread local variables are in the Java heap in a regular array, i.e., array write barriers are needed.

Reply:
So the barrier kind is really a property of VMThreadLocalInfo?

Reply:
No, the barrier kind is a property of how the thread local storage is implemented. Which is all encapsulated in this class, so defining the barrier kind in this class is correct.

Reply:
Sounds good.